### PR TITLE
Avoid pulling same images multiple times ⚡️

### PR DESCRIPTION
**Pull request by KoditkarVedant**
**What I did**
`docker compose pull` pulls images even if they are present locally and if there are multiple services using the same image it pulls them multiple times. This pull request avoid pulling image if it is already present and only fetch image once if required by many services.

**Related issue**
#8768 

**(not mandatory) A picture of a cute animal, if possible in relation with what you did**
image

### CODE REVIEW

**Review by ndeloof — APPROVED** [review]

**Comment by KoditkarVedant** [comment]
@ndeloof I can not re-run the failed tests and they are also failing on `V2` branch when I run them locally. Can you please look into this? Do let me know in case I can help.

**Comment by KoditkarVedant** [comment]
@ndeloof need your help to merge this pull request.

**Comment by ndeloof** [comment]
need TestComposePull to be fixed

**Comment by KoditkarVedant** [comment]
oh, I thought it may be because of some tests which fails sometime. I will fix it.

**Comment by KoditkarVedant** [comment]
@ndeloof I tried running the test case on my local it ran without any error.

image

Can we rerun the integration pipeline if it gets resolved?

**Comment by glours** [comment]
@KoditkarVedant I just restartes the CI, the `TestComposePull` test is still in error

**Comment by KoditkarVedant** [comment]
@glours I've merged the latest changes from V2 in my branch. Lets see if the test still fails. No idea why I'm not able to reproduce this locally.

**Comment by KoditkarVedant** [comment]
Finally, Test is failing locally. I will fix it.

**Comment by KoditkarVedant** [comment]
@ndeloof @glours Fixed the TestComposePull.

**Comment by n-rodriguez** [comment]
Hi there! Any news?

**Comment by KoditkarVedant** [comment]
@n-rodriguez this has been delayed for long time. I need to get my local setup up and running again. I will spend some time over the weekend.

**Comment by KoditkarVedant** [comment]
@ndeloof Need help to get this PR over line.

File: pkg/e2e/compose_test.go [hunk 1]
```diff
@@ -125,8 +125,28 @@ func TestComposePull(t *testing.T) {
...
+		"Skipped - Image is already present locally",
+		"Skipped - Image is already being pulled by",
+		"simple Pulled",
+		"another Pulled",
+	}
+
+	assert.Assert(t, contains(output, expected))
+}
+
+func contains(str string, array []string) bool {
```
Comment (laurazard) [inline pkg/e2e/compose_test.go:139]: nit: `contains()` could be a bit cleaner:
```
func contains(str string, array []string) bool {
	for _, val := range array {
		if strings.Contains(str, val) {
			return true
		}
	}
	return false
}
```

File: pkg/e2e/compose_test.go [hunk 2]
```diff
@@ -125,8 +125,28 @@ func TestComposePull(t *testing.T) {
 	res := c.RunDockerComposeCmd(t, "--project-directory", "fixtures/simple-composefile", "pull")
 	output := res.Combined()
 
-	assert.Assert(t, strings.Contains(output, "simple Pulled"))
-	assert.Assert(t, strings.Contains(output, "another Pulled"))
+	expected := []string{
```
Comment (laurazard) [inline pkg/e2e/compose_test.go:128]: nit: could we make this a bit more deterministic? Playing around with it locally, seems like `Skipped - Image is already being pulled by` is the only `Skipped - xxx` message that needs checking, and then maybe `Pulled`. Maybe we could do away with `contains()` entirely and just check the output as a whole contains those key phrases

Comment (KoditkarVedant) [reply pkg/e2e/compose_test.go:128]: I believe I have an idea how to make this more explicit

Comment (KoditkarVedant) [reply pkg/e2e/compose_test.go:128]: done

File: pkg/e2e/compose_test.go (hunk 1 above)
Comment (KoditkarVedant) [reply pkg/e2e/compose_test.go:139]: after updating test this is no longer required

**Comment by KoditkarVedant** [comment]
@laurazard done with the changes

**Comment by laurazard** [comment]
@KoditkarVedant looks like there are some CI failures. I can’t really look into it (on mobile) can you take a look and fix it?

The changes to the tests are great ☺️

**Comment by KoditkarVedant** [comment]
@laurazard I have changed the image used for that to avoid any race condition while running the test case. Lets see if that resolves the issue. I do not get any error for TestComposePull when I run it locally.

**Comment by KoditkarVedant** [comment]
@laurazard I tried multiple ways but this tests doesn't seem to pass on the CI server. It works as expected on local. can you please have a look at this? 

`Failing test: TestComposePull/Verify_skipped_pull_if_image_is_already_present_locally`
My idea is to pull the required image first so when I try to pull it using the compose tool it will already be present locally.

**Comment by laurazard** [comment]
I'll take a look at it :)

**Comment by laurazard** [comment]
@KoditkarVedant I've been looking at this locally and I can consistently replicate the error. Moreover, I was trying to figure out where the error was coming from and searching all over the project, I could not find where the `... Image is already present locally` message is supposed to come from. Did you forget to commit something?

**Comment by KoditkarVedant** [comment]
@laurazard Thank you - I see what I was doing wrong for so long. I was using the old docker-compose build that I build previously so  rebuilding the CLI starts failing the tests locally. I have added the fix.

**Comment by laurazard** [comment]
awesome @KoditkarVedant! We'll try to get it reviewed and merged soon :)

**Review by nicksieger — APPROVED** [review]
LGTM! Thanks!

**Review by laurazard — APPROVED** [review]
LGTM! Thanks :)

**Comment by geozeke** [comment]
Thanks for the hard work on this project!

In this version it looks like `docker-compose pull` will now not pull the latest updates to an image. For example, if my `docker-compose.yml` file contains this: `image: mariadb:latest`, then even if there's an updated `mariadb` image, `docker-compose` reports:

```shell
mariadb Skipped - Image is already present locally
```

**Comment by hugalafutro** [comment]
I can confirm the same, none of my updated :latest images download, claiming they exist locally. Downgrading to compose 2.9.0 updates local images correctly when running `docker compose pull`.

**Comment by geozeke** [comment]
Same for me, @hugalafutro. I didn't see a documented change in behavior, but `docker-compose pull` has always worked for me the way you describe.

**Comment by martadinata666** [comment]
So related to the issue that this PR want to solve, seems this already solved long ago, i tested a compose file with same `image` it will only pull once and the other will wait instead redownload again.
```
docker compose pull    
[+] Running 13/13
 ⠿ db2 Pulled                                                                                                70.3s
 ⠿ db Pulled                                                                                                 70.3s
   ⠿ a749a280e3e9 Pull complete                                                                              24.3s
   ⠿ 3f7d70702fc5 Pull complete                                                                              24.7s
   ⠿ b9ffe290feb5 Pull complete                                                                              25.9s
   ⠿ 5a430076265b Pull complete                                                                              26.6s
   ⠿ 45c880491c15 Pull complete                                                                              26.8s
   ⠿ 41ca8382cebf Pull complete                                                                              28.1s
   ⠿ 88502488cc76 Pull complete                                                                              28.4s
   ⠿ 8ae7307cb02f Pull complete                                                                              28.7s
   ⠿ df60df6710cc Pull complete                                                                              63.8s
   ⠿ bb82d118ee8f Pull complete                                                                              64.0s
   ⠿ 6ce2f3952ce5 Pull complete  
```

the compose file
```
version: "3.8"
services:
 db:
    image: mariadb:10.6
    environment:
      - MARIADB_AUTO_UPGRADE=1
      - MARIADB_ROOT_PASSWORD=rootpass
      - TZ=Asia/Jakarta
      - MARIADB_DATABASE=test
      - MARIADB_USER=testuser
      - MARIADB_PASSWORD=testpass
    volumes:
      - ./db:/var/lib/mysql

 db2:
    image: mariadb:10.6
    environment:
      - MARIADB_AUTO_UPGRADE=1
      - MARIADB_ROOT_PASSWORD=rootpass
      - TZ=Asia/Jakarta
      - MARIADB_DATABASE=test
      - MARIADB_USER=testuser
      - MARIADB_PASSWORD=testpass
    volumes:
      - ./db2:/var/lib/mysql

```

**Comment by geozeke** [comment]
Looks like there's a new issue open for this: #9773

**Pull request merged**